File: html5/src/chibi/ChibiBuilder.js
// -----------------------------------------------------------------------------
// ChibiBuilder — a recipe becomes a subtree of primitive-mesh actors.
//
// The whole feature rests on one observation: a chibi's proportions hide its
// joints, so nothing has to bend, so nothing needs a skinning shader, a bone
// palette or a model file. A character is a hierarchy of cubes, spheres and
// capsules, which is exactly what both engines can already draw.
//
// A joint actor carries rotation only; the parts hanging off it are separate
// children holding the offset and the scale. That is what puts a limb's pivot at
// the shoulder rather than half-way down the upper arm, and it means animation
// writes nothing but `localEulerAngles` on sixteen actors.
//
// Mirrored in SexyBiscuit.Engine/Chibi/ChibiBuilder.cs.
// -----------------------------------------------------------------------------

import { Actor } from '../core/Actor.js';
import { Transform3D } from '../core/Transform3D.js';
import { MeshRenderer } from '../rendering/MeshRenderer.js';
import { MeshPrimitive } from '../rendering/PrimitiveMesh.js';
import { Vector3, Color } from '../math/index.js';
import {
    ChibiParts, JOINTS, SOCKETS, SLOTS, ROOT_NAME, isMirrored, expand, mirrorVector, mirrorEuler,
} from './ChibiRig.js';
import { normalise } from './ChibiRecipe.js';

/** Chibis are flat-shaded toys, not car paint. */
const ROUGHNESS = 0.78;

/** Slots authored against the reference head, and so scaled onto the real one. */
const FITTED_TO_HEAD = new Set(['hair', 'eyes']);

/** Sockets on the head, which have to move with it for the same reason. */
const FITTED_SOCKETS = new Set(['Head', 'Face']);

const NO_FIT = [1, 1, 1];

/** One built character: the actor plus the lookups a game and the panel need. */
export class ChibiBuild {
    constructor(actor, recipe) {
        /** @type {Actor} The root. Move this, not the parts. */
        this.actor = actor;
        /** @type {object} The recipe as it was normalised, not as it was passed. */
        this.recipe = recipe;
        /** @type {Map<string, Transform3D>} Joint name to its transform, resolved once. */
        this.joints = new Map();
        /** @type {Map<string, Actor>} Socket name (without the prefix) to its actor. */
        this.sockets = new Map();
        /** @type {Array<{renderer: MeshRenderer, colour: string}>} Parts, by colour slot. */
        this.parts = [];
    }

    /** Repaints every part drawn in one colour slot. */
    setColour(slot, hex) {
        if (!(slot in this.recipe.colours)) return false;
        this.recipe.colours[slot] = hex;
        const colour = Color.from(hex);
        for (const part of this.parts) {
            if (part.colour === slot) part.renderer.albedoColor = colour;
        }
        return true;
    }
}

function proportion(recipe, name) {
    if (!name) return 1;
    return recipe.proportions[name] ?? 1;
}

function makeActor(name, parent) {
    const actor = new Actor(name);
    actor.addComponent(Transform3D);
    // false, not the default: a rig's offsets are already local, so keeping the
    // world transform would drag every joint back to where it was standing.
    if (parent) actor.attachTo(parent, false);
    return actor;
}

/**
 * Adds one primitive piece under `joint`, returning its renderer so the caller
 * can recolour it later without walking the tree again.
 */
function addPiece(piece, jointActor, side, recipe, build, name, colourOverride = null,
                  fit = NO_FIT) {
    const width = proportion(recipe, piece.widthScale);
    const length = proportion(recipe, piece.lengthScale);

    const actor = makeActor(name, jointActor);
    const t = actor.getComponent(Transform3D);

    const pos = mirrorVector(piece.pos, side);
    t.localPosition = new Vector3(
        pos[0] * width * fit[0], pos[1] * length * fit[1], pos[2] * width * fit[2]);
    t.localScale = new Vector3(
        piece.scale[0] * width * fit[0],
        piece.scale[1] * length * fit[1],
        piece.scale[2] * width * fit[2]);
    if (piece.rot) t.localEulerAngles = Vector3.from(mirrorEuler(piece.rot, side));

    const renderer = actor.addComponent(MeshRenderer);
    renderer.setPrimitive(MeshPrimitive[piece.mesh] ?? MeshPrimitive.Cube);
    renderer.albedoColor = Color.from(colourOverride ?? recipe.colours[piece.colour] ?? '#FFFFFF');
    renderer.roughness = ROUGHNESS;

    // An overridden accessory colour is not in any slot, so setColour must not
    // repaint it later when the slot it borrowed its name from changes.
    build.parts.push({ renderer, colour: colourOverride ? null : piece.colour });
    return actor;
}

/**
 * Builds a character.
 *
 * @param {object} recipe A recipe, raw or already normalised.
 * @param {(message: string) => void} [onWarning] Told about anything dropped.
 * @returns {ChibiBuild}
 */
export function build(recipe, onWarning = null) {
    const r = normalise(recipe, onWarning);

    const root = new Actor(r.name || ROOT_NAME);
    root.addComponent(Transform3D).localScale = new Vector3(
        r.proportions.height, r.proportions.height, r.proportions.height);

    const result = new ChibiBuild(root, r);
    const actors = new Map([['', root]]);
    const headFit = ChibiParts.headFit?.[r.style.head] ?? NO_FIT;
    // Eyes get their own fit: it carries how far forward that head's face is, which
    // differs by shape in a way the skull's proportions alone do not describe.
    const eyeFit = ChibiParts.headEyeFit?.[r.style.head] ?? headFit;

    // ---- Joints, parents first so every attach finds its parent -------------
    for (const joint of JOINTS) {
        const parent = actors.get(joint.parent) ?? root;
        const actor = makeActor(joint.name, parent);
        const scale = proportion(r, joint.posScale);
        actor.getComponent(Transform3D).localPosition = new Vector3(
            joint.pos[0], joint.pos[1] * scale, joint.pos[2]);

        actors.set(joint.name, actor);
        result.joints.set(joint.name, actor.getComponent(Transform3D));
    }

    // ---- Sockets: empty actors a game hangs a hat or a sword from -----------
    for (const socket of SOCKETS) {
        const actor = makeActor(`Socket_${socket.name}`, actors.get(socket.joint) ?? root);
        const fit = FITTED_SOCKETS.has(socket.name) ? headFit : NO_FIT;
        actor.getComponent(Transform3D).localPosition = new Vector3(
            socket.pos[0] * fit[0], socket.pos[1] * fit[1], socket.pos[2] * fit[2]);
        result.sockets.set(socket.name, actor);
    }

    // ---- The parts themselves ----------------------------------------------
    for (const slot of SLOTS) {
        const pieces = ChibiParts.parts[slot]?.[r.style[slot]] ?? [];
        const fit = slot === 'eyes' ? eyeFit : (FITTED_TO_HEAD.has(slot) ? headFit : NO_FIT);
        for (const piece of pieces) {
            for (const side of isMirrored(piece.joint) ? ['L', 'R'] : [null]) {
                const jointName = side ? expand(piece.joint, side) : piece.joint;
                const jointActor = actors.get(jointName);
                if (!jointActor) {
                    if (onWarning) onWarning(`${slot}/${r.style[slot]} names joint '${jointName}', which the rig has not got.`);
                    continue;
                }
                addPiece(piece, jointActor, side ?? 'L', r, result,
                    `${slot}_${piece.mesh}`, null, fit);
            }
        }
    }

    // ---- Accessories, which hang off sockets rather than joints -------------
    for (const entry of r.accessories) {
        for (const piece of ChibiParts.accessories[entry.part] ?? []) {
            for (const side of isMirrored(piece.socket) ? ['L', 'R'] : [null]) {
                const name = side ? expand(piece.socket, side) : piece.socket;
                const socket = result.sockets.get(name);
                if (!socket) {
                    if (onWarning) onWarning(`accessory '${entry.part}' names socket '${name}', which the rig has not got.`);
                    continue;
                }
                // A hat is placed relative to the socket, which has already moved
                // with the head, so only its size needs fitting -- and only on the
                // head, where a mismatch is a hat that does not sit on the skull.
                const fit = FITTED_SOCKETS.has(name) ? headFit : NO_FIT;
                addPiece(piece, socket, side ?? 'L', r, result,
                    `${entry.part}_${piece.mesh}`, entry.colour, fit);
            }
        }
    }

    return result;
}

File: SexyBiscuit.Engine/Chibi/ChibiBuilder.cs
using Microsoft.Xna.Framework;
using SexyBiscuit.Engine.Core;
using SexyBiscuit.Engine.Rendering;

namespace SexyBiscuit.Engine.Chibi;

/// <summary>One built character: the actor, plus the lookups a game and a panel need.</summary>
public sealed class ChibiBuild
{
    /// <summary>The root. Move this, not the parts.</summary>
    public Actor Actor { get; }

    /// <summary>The recipe as it was normalised, not as it was passed.</summary>
    public ChibiRecipe Recipe { get; }

    /// <summary>Joint name to its transform, resolved once so nothing searches per frame.</summary>
    public Dictionary<string, Transform3D> Joints { get; } = new(StringComparer.Ordinal);

    /// <summary>Socket name, without the <c>Socket_</c> prefix its actor carries, to that actor.</summary>
    public Dictionary<string, Actor> Sockets { get; } = new(StringComparer.Ordinal);

    /// <summary>Every drawn piece and the colour slot it takes its colour from.</summary>
    public List<(MeshRenderer Renderer, string? Colour)> Parts { get; } = new();

    internal ChibiBuild(Actor actor, ChibiRecipe recipe)
    {
        Actor = actor;
        Recipe = recipe;
    }

    /// <summary>Repaints every part drawn in one colour slot. False for an unknown slot.</summary>
    public bool SetColour(string slot, string hex)
    {
        if (!Recipe.Colours.ContainsKey(slot)) return false;

        Recipe.Colours[slot] = hex;
        Color colour = ChibiBuilder.ParseColour(hex);
        foreach (var (renderer, name) in Parts)
        {
            if (name == slot) renderer.AlbedoColor = colour;
        }
        return true;
    }
}

/// <summary>
/// A recipe becomes a subtree of primitive-mesh actors.
/// </summary>
/// <remarks>
/// <para>
/// The whole feature rests on one observation: a chibi's proportions hide its joints, so
/// nothing has to bend, so nothing needs a skinning shader, a bone palette or a model file.
/// A character is a hierarchy of cubes, spheres and capsules, which is what both engines
/// can already draw.
/// </para>
/// <para>
/// A joint actor carries rotation only; the parts hanging off it are separate children
/// holding the offset and the scale. That is what puts a limb's pivot at the shoulder
/// rather than half-way down the upper arm, and it means animation writes nothing but
/// <see cref="Transform3D.LocalEulerAngles"/> on sixteen actors.
/// </para>
/// <para>Mirrored in <c>html5/src/chibi/ChibiBuilder.js</c>.</para>
/// </remarks>
public static class ChibiBuilder
{
    /// <summary>Chibis are flat-shaded toys, not car paint.</summary>
    private const float Roughness = 0.78f;

    /// <summary>Slots authored against the reference head, and so scaled onto the real one.</summary>
    private static readonly HashSet<string> FittedToHead = new(StringComparer.Ordinal) { "hair", "eyes" };

    /// <summary>Sockets on the head, which have to move with it for the same reason.</summary>
    private static readonly HashSet<string> FittedSockets = new(StringComparer.Ordinal) { "Head", "Face" };

    /// <summary>Parses <c>#RRGGBB</c> or <c>#RRGGBBAA</c>. Anything else is white.</summary>
    internal static Color ParseColour(string? value)
    {
        string hex = (value ?? string.Empty).TrimStart('#');
        if (hex.Length < 6) return Color.White;

        try
        {
            byte Channel(int i) => Convert.ToByte(hex.Substring(i, 2), 16);
            return new Color(Channel(0), Channel(2), Channel(4), hex.Length >= 8 ? Channel(6) : (byte)255);
        }
        catch (FormatException) { return Color.White; }
    }

    /// <summary>Builds a character.</summary>
    /// <param name="recipe">A recipe, already normalised.</param>
    /// <param name="onWarning">Told about anything the part table could not place.</param>
    public static ChibiBuild Build(ChibiRecipe recipe, Action<string>? onWarning = null)
    {
        float height = recipe.Proportion("height");

        var root = new Actor(string.IsNullOrWhiteSpace(recipe.Name) ? "Chibi" : recipe.Name);
        root.AddComponent<Transform3D>().LocalScale = new Vector3(height, height, height);

        var build = new ChibiBuild(root, recipe);
        var actors = new Dictionary<string, Actor>(StringComparer.Ordinal) { [string.Empty] = root };

        Vector3 headFit = ChibiParts.HeadFit(recipe.Style.GetValueOrDefault("head", "Round"));
        // Eyes get their own fit: it carries how far forward that head's face is, which
        // differs by shape in a way the skull's proportions alone do not describe.
        Vector3 eyeFit = ChibiParts.HeadEyeFit(recipe.Style.GetValueOrDefault("head", "Round"));

        // ---- Joints, parents first so every attach finds its parent -----------------
        foreach (ChibiJoint joint in ChibiParts.Joints)
        {
            Actor parent = actors.GetValueOrDefault(joint.Parent, root);
            Actor actor = MakeActor(joint.Name, parent);

            float scale = recipe.Proportion(joint.PosScale);
            actor.GetComponent<Transform3D>()!.LocalPosition =
                new Vector3(joint.Pos.X, joint.Pos.Y * scale, joint.Pos.Z);

            actors[joint.Name] = actor;
            build.Joints[joint.Name] = actor.GetComponent<Transform3D>()!;
        }

        // ---- Sockets: empty actors a game hangs a hat or a sword from ---------------
        foreach (ChibiSocket socket in ChibiParts.Sockets)
        {
            Actor actor = MakeActor($"Socket_{socket.Name}", actors.GetValueOrDefault(socket.Joint, root));
            Vector3 fit = FittedSockets.Contains(socket.Name) ? headFit : Vector3.One;
            actor.GetComponent<Transform3D>()!.LocalPosition = socket.Pos * fit;
            build.Sockets[socket.Name] = actor;
        }

        // ---- The parts themselves ---------------------------------------------------
        foreach (string slot in ChibiParts.Slots)
        {
            string variant = recipe.Style.GetValueOrDefault(slot, string.Empty);
            Vector3 fit = slot == "eyes" ? eyeFit
                : FittedToHead.Contains(slot) ? headFit : Vector3.One;

            foreach (ChibiPiece piece in ChibiParts.Pieces(slot, variant))
            {
                foreach (string side in ChibiParts.SidesOf(piece.Joint))
                {
                    string jointName = ChibiParts.Expand(piece.Joint, side);
                    if (!actors.TryGetValue(jointName, out Actor? jointActor))
                    {
                        onWarning?.Invoke(
                            $"{slot}/{variant} names joint '{jointName}', which the rig has not got.");
                        continue;
                    }
                    AddPiece(piece, jointActor, side, recipe, build, $"{slot}_{piece.Mesh}", null, fit);
                }
            }
        }

        // ---- Accessories, which hang off sockets rather than joints ------------------
        foreach (ChibiRecipe.Accessory entry in recipe.Accessories)
        {
            foreach (ChibiPiece piece in ChibiParts.Accessory(entry.Part))
            {
                foreach (string side in ChibiParts.SidesOf(piece.Socket))
                {
                    string name = ChibiParts.Expand(piece.Socket, side);
                    if (!build.Sockets.TryGetValue(name, out Actor? socket))
                    {
                        onWarning?.Invoke(
                            $"accessory '{entry.Part}' names socket '{name}', which the rig has not got.");
                        continue;
                    }

                    // A hat is placed relative to the socket, which has already moved with
                    // the head, so only its size needs fitting -- and only on the head,
                    // where a mismatch is a hat that does not sit on the skull.
                    Vector3 fit = FittedSockets.Contains(name) ? headFit : Vector3.One;
                    AddPiece(piece, socket, side, recipe, build,
                        $"{entry.Part}_{piece.Mesh}", entry.Colour, fit);
                }
            }
        }

        return build;
    }

    private static Actor MakeActor(string name, Actor? parent)
    {
        var actor = new Actor(name);
        actor.AddComponent<Transform3D>();
        // false, not the default: a rig's offsets are already local, so keeping the world
        // transform would drag every joint back to where it was standing.
        if (parent != null) actor.AttachTo(parent, keepWorldTransform: false);
        return actor;
    }

    private static void AddPiece(ChibiPiece piece, Actor jointActor, string side, ChibiRecipe recipe,
                                 ChibiBuild build, string name, string? colourOverride, Vector3 fit)
    {
        float width = recipe.Proportion(piece.WidthScale);
        float length = recipe.Proportion(piece.LengthScale);

        Actor actor = MakeActor(name, jointActor);
        var transform = actor.GetComponent<Transform3D>()!;

        Vector3 pos = ChibiParts.MirrorVector(piece.Pos, side);
        transform.LocalPosition = new Vector3(
            pos.X * width * fit.X, pos.Y * length * fit.Y, pos.Z * width * fit.Z);
        transform.LocalScale = new Vector3(
            piece.Scale.X * width * fit.X, piece.Scale.Y * length * fit.Y, piece.Scale.Z * width * fit.Z);
        if (piece.Rot != Vector3.Zero)
            transform.LocalEulerAngles = ChibiParts.MirrorEuler(piece.Rot, side);

        var renderer = actor.AddComponent<MeshRenderer>();
        renderer.MeshType = Enum.TryParse(piece.Mesh, out MeshPrimitive primitive)
            ? primitive : MeshPrimitive.Cube;
        renderer.AlbedoColor = ParseColour(
            colourOverride ?? recipe.Colours.GetValueOrDefault(piece.Colour, "#FFFFFF"));
        renderer.Roughness = Roughness;

        // An overridden accessory colour is not in any slot, so SetColour must not repaint
        // it later when the slot it borrowed its name from changes.
        build.Parts.Add((renderer, colourOverride != null ? null : piece.Colour));
    }
}
